File: 15.3-sum.js
/*
 * @lc app=leetcode id=15 lang=javascript
 *
 * [15] 3Sum
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
  nums.sort((a, b) => a - b);
  const len = nums.length;
  const res = new Set();

  for (let i = 0; i < len - 2; ++i) {
    let j = i + 1,
      k = len - 1;
    while (j < k) {
      const [a, b, c] = [nums[i], nums[j], nums[k]];
      const sum = a + b + c;
      if (sum < 0) ++j;
      else if (sum > 0) --k;
      else {
        res.add(`${a},${b},${c}`);
        ++j;
        --k;
      }
    }
  }
  return [...res].map((str) => str.split(","));
};
// @lc code=end
const res = threeSum([-1, 0, 1, 2, -1, -4]);
console.log(res);

File: 15.3-sum.rb
#
# @lc app=leetcode id=15 lang=ruby
#
# [15] 3Sum
#

# @lc code=start
# @param {Integer[]} nums
# @return {Integer[][]}
def three_sum(nums)
  nums.sort!
  ans = Set.new
  (0..nums.length - 3).each do |i|
    j = i + 1
    k = nums.length - 1
    while j < k
      sum = nums[i] + nums[j] + nums[k]
      if sum < 0
        j += 1
      elsif sum > 0
        k -= 1
      else
        ans << [nums[i], nums[j], nums[k]]
        j += 1
        k -= 1
      end
    end
  end
  [*ans]
end
# @lc code=end

three_sum([-1, 0, 1, 2, -1, -4])
